src/OpenTelemetry/Metrics/MetricPoint.cs

```diff
@@ -1310,34 +1311,20 @@ internal void TakeSnapshotWithExemplar(bool outputDelta)
 
                     this.MetricPointStatus = MetricPointStatus.NoCollectPending;
 
-                    ReleaseLock(ref histogram.IsCriticalSectionOccupied);
+                    this.mpComponents.ReleaseLock();
 
                     break;
                 }
         }
     }
 
-    private static void AcquireLock(ref int isCriticalSectionOccupied)
-    {
-        var sw = default(SpinWait);
-        while (Interlocked.Exchange(ref isCriticalSectionOccupied, 1) != 0)
```

Comment:
`Interlocked.Exchange` here seems totally wrong. Am I missing something or do we have a big bug on our hands?

Reply:
This also exposes a big gap the tests -- we do have multi-threading test, and it is passing despite it.

Reply:
A thread will keep spinning until it confirms that the original value here was 0 (i.e it was unoccupied..)... The thread originally holding this, would have to set it to 0, for the other thread to enter, and only one will win/enter...
Seems okay ?

I need to think more...

Reply:
Writing down my thinking:

This code will set criticalSection to 1 always. if the original value was 0, it assumes it won the lock and exits. If not, it spins and retries indefinitely until the value is originally 0.

Reply:
Thinking about it more, logically it is actually fine. Sorry for the false alarm 😄 I perf tested `Exchange` vs `CompareExchange` and the former is faster so I put it back.